src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp

```diff
@@ -1083,25 +1083,25 @@ void BinarySpaceTree<MetricType, StatisticType, MatType, BoundType, SplitType>::
     right = NULL;
   }
 
-  ar & BOOST_SERIALIZATION_NVP(begin);
-  ar & BOOST_SERIALIZATION_NVP(count);
-  ar & BOOST_SERIALIZATION_NVP(bound);
-  ar & BOOST_SERIALIZATION_NVP(stat);
+  ar & CEREAL_NVP(begin);
+  ar & CEREAL_NVP(count);
+  ar & CEREAL_NVP(bound);
+  ar & CEREAL_NVP(stat);
 
-  ar & BOOST_SERIALIZATION_NVP(parentDistance);
-  ar & BOOST_SERIALIZATION_NVP(furthestDescendantDistance);
-  ar & BOOST_SERIALIZATION_NVP(dataset);
+  ar & CEREAL_NVP(parentDistance);
+  ar & CEREAL_NVP(furthestDescendantDistance);
+  ar & CEREAL_NVP(dataset);
 
-  // Save children last; otherwise boost::serialization gets confused.
+  // Save children last; otherwise cereal gets confused.
   bool hasLeft = (left != NULL);
   bool hasRight = (right != NULL);
 
-  ar & BOOST_SERIALIZATION_NVP(hasLeft);
-  ar & BOOST_SERIALIZATION_NVP(hasRight);
+  ar & CEREAL_NVP(hasLeft);
+  ar & CEREAL_NVP(hasRight);
   if (hasLeft)
-    ar & BOOST_SERIALIZATION_NVP(left);
+    ar & CEREAL_NVP(left);
   if (hasRight)
-    ar & BOOST_SERIALIZATION_NVP(right);
+    ar & CEREAL_NVP(right);
 
```

Comment:
Another idea (added as a separate comment to split the discussion) would be to make some kind of nice wrapper function that can serialize raw pointers, so that we could make calls like this:

```
ar & CEREAL_POINTER_WRAPPER(this->left);
ar & CEREAL_POINTER_WRAPPER(this->right);
```

Basically `CEREAL_POINTER_WRAPPER()` would take in a `T*` (for whatever type `T` is), and then do some "magic" to actually serialize it.  Maybe the following code is helpful for thinking about what that "magic" could be:

```
  // for saving
  std::unique_ptr<TreeType> leftUnique(*this->left); // call copy constructor...
  ar & BOOST_SERIALIZATION_NVP(leftUnique);

  // for loading
  std::unique_ptr<TreeType> leftUnique;
  ar & BOOST_SERIALIZATION_NVP(leftUnique); // deserialization
  this->left = new TreeType(std::move(*leftUnique)); // use move constructor
```

That saving code is really bad though; calling the copy constructor is painful!  But we can't call the move constructor there, since we still need the object to be intact after serialization.  Maybe a `weak_ptr<>` or something we can wrap around a raw pointer might work better?  For loading, using the move constructor after cereal constructs the object hopefully shouldn't be *too* computationally expensive.  It's at least the best thing I can think of.  But maybe there is some clever way we can manage to get cereal to give a smart pointer back, and safely "convert" this to a raw pointer.